### PR TITLE
Consolidate filter status information into a new component, `FilterStatus`

### PR DESCRIPTION
**Pull request by lyzadanger**
This PR consolidates filter status information in the client into a single component, `FilterStatus` (behind feature flag) and fixes confusing behavior arising from combining selected annotations and queries (not behind feature flag).

This PR takes over where https://github.com/hypothesis/client/pull/2383 left off, and makes the changes outlined [in this document](https://docs.google.com/document/d/1I_jT2hVXwWVtMNWqcC1mxQpmckLvu2SD_d2A0Reo0Us/edit#) (requires Hypothesis team access). 

Fixes https://github.com/hypothesis/client/issues/2270
Fixes https://github.com/hypothesis/client/issues/1383 (behind feature flag)

Putting into draft until I have a chance to integrate initial feedback from @robertknight on #2383  — this PR is cleaned up and has tests, however.

### CODE REVIEW

File: src/sidebar/components/filter-status.js [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+ * state:
+ * - selection: One or more annotations have been selected in the document by
+ *   the user, or we're looking at a direct-linked annotation. Button clears
+ *   the selection. This mode supersedes all others (i.e. other filters ignored).
+ * - query: There is a user-entered filter query. It can "stack" with user-focus.
+ *    Button clears the selection (query, etc.), but will leave user-focus mode
+ *    intact.
+ * - focus: Neither of the above two modes apply and user-focus mode is
+ *   configured (it may or may not be active). Button toggles activation of
+ *   user-focus mode.
```
Comment (robertknight) [inline src/sidebar/components/filter-status.js:26]: It sounds from reading this like "focus" is not mutually-exclusive vs. the other modes.

File: src/sidebar/components/filter-status.js [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+ * Render information about the currently-applied filters and allow the
+ * user to clear filters or toggle user-focus mode on and off
+ *
+ * When any filtering is applied, we are in one of 3 mutually-exclusive filtering
+ * "modes". In any mode, we display descriptive text about what the user
+ * is looking at, and present one button to take action on the current filtering
+ * state:
+ * - selection: One or more annotations have been selected in the document by
+ *   the user, or we're looking at a direct-linked annotation. Button clears
+ *   the selection. This mode supersedes all others (i.e. other filters ignored).
```
Comment (robertknight) [inline src/sidebar/components/filter-status.js:20]: The comment says "this mode supersedes all others" but in a `switch` block below there is code that changes the text in `selection` mode depending on whether focus mode is active.

File: src/sidebar/components/filter-status.js [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+          : `Show only ${filterState.focusDisplayName}`;
+      } else {
+        // When user focus is applied and there are some forced-visible threads,
+        // this special case applies. The button will clear force-visible threads
+        // but leave user focus intact
+        buttonText = 'Reset filters';
+      }
+      break;
+    case 'query':
+    default:
```
Comment (robertknight) [inline src/sidebar/components/filter-status.js:96]: If `filterMode` has no possibilities other than `selection`, `focus` or `query` then you don't need the `default` case here.

File: src/sidebar/components/search-input.js [hunk 4]
```diff
@@ -29,7 +29,11 @@ export default function SearchInput({ alwaysExpanded, query, onSearch }) {
 
   const onSubmit = e => {
     e.preventDefault();
-    onSearch(input.current.value);
+    if (input.current.value || prevQuery) {
+      // Don't set an initial empty query, but allow a later empty query to
+      // clear `prevQuery`
+      onSearch(input.current.value);
+    }
```
Comment (robertknight) [inline src/sidebar/components/search-input.js:36]: Is this change something that could be separated out of this PR?

File: src/sidebar/components/sidebar-content.js [hunk 5]
```diff
@@ -45,6 +46,9 @@ function SidebarContent({
     ? tabForAnnotation(linkedAnnotation)
     : null;
   const searchUris = useStore(store => store.searchUris());
+  const filterStatusEnabled = useStore(store =>
+    store.isFeatureEnabled('client_filter_status')
```
Comment (robertknight) [inline src/sidebar/components/sidebar-content.js:50]: 👍

File: src/sidebar/components/test/filter-status-test.js [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+          createComponent(),
+          "Showing 1 result for 'biscuits' by Ebenezer Studentolog (and 2 more)"
+        );
+      });
+
+      it('should provide a "Clear search" button', () => {
+        assertButton(createComponent(), {
+          text: 'Clear search',
+          icon: 'cancel',
+          callback: fakeStore.clearSelection,
```
Comment (robertknight) [inline src/sidebar/components/test/filter-status-test.js:285]: This test is repeated a few times. Might be worth refactoring into a helper.

File: src/sidebar/components/test/filter-status-test.js [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+
+    beforeEach(() => {
+      filterState = {
+        ...getFilterState(),
+        focusActive: true,
+        focusConfigured: true,
+        focusDisplayName: 'Ebenezer Studentolog',
+        selectedCount: 2,
+      };
+      fakeStore.filterState.returns(filterState);
```
Comment (robertknight) [inline src/sidebar/components/test/filter-status-test.js:302]: It may be helpful to factor out a helper which maps the logical state ("user focus mode active") to the specific setup logic that is needed. This will reduce the amount of work required if changes are made to how eg. focus mode is represented in future.

File: src/sidebar/components/test/search-input-test.js [hunk 8]
```diff
@@ -61,6 +61,25 @@ describe('SearchInput', () => {
     assert.calledWith(onSearch, 'new-query');
   });
 
+  it('does not set an initial empty query when form is submitted', () => {
+    // If the first query entered is empty, it will be ignored
+    const onSearch = sinon.stub();
+    const wrapper = createSearchInput({ onSearch });
+    typeQuery(wrapper, '');
+    wrapper.find('form').simulate('submit');
+    assert.notCalled(onSearch);
```
Comment (robertknight) [inline src/sidebar/components/test/search-input-test.js:70]: Can you remind me why the change in `SearchInput`'s behavior was needed?

File: src/sidebar/services/threads.js [hunk 9]
```diff
@@ -1,15 +1,24 @@
...
+   *
+   * @param {Thread} thread
    */
   function forceVisible(thread) {
     thread.children.forEach(child => {
       forceVisible(child);
     });
-    store.setForcedVisible(thread.id, true);
+    if (!thread.visible) {
+      store.setForcedVisible(thread.id, true);
```
Comment (robertknight) [inline src/sidebar/services/threads.js:20]: Is this just an optimization?

File: src/sidebar/store/modules/selection.js [hunk 10]
```diff
@@ -163,27 +173,38 @@ const setTab = (newTab, oldTab) => {
   };
 };
 
+/**
+ * Return state object representing a reset selection. This
+ * resets user-set filters (but leaves focus mode intact)
+ *
+ * @return {Object}
```
Comment (robertknight) [inline src/sidebar/store/modules/selection.js:180]: Was there a reason for the `@return {Object}` here? Omitting it doesn't cause an issue AFAICS and it will allow TS to do better inference in future.

File: src/sidebar/store/modules/selection.js [hunk 11]
```diff
@@ -163,27 +173,38 @@ const setTab = (newTab, oldTab) => {
...
+ */
+const resetSelection = () => {
```
Comment (robertknight) [inline src/sidebar/store/modules/selection.js:182]: Given that store state is immutable, this can just be a constant. You could wrap the literal with the `immutable` helper if you wanted to guard against accidental mutation.

File: src/sidebar/store/modules/test/selection-test.js [hunk 12]
```diff
@@ -64,9 +64,11 @@ describe('sidebar/store/modules/selection', () => {
...
-    it('returns true if user-focused mode applied', () => {
-      store = createStore([selection], [{ focus: { user: {} } }]);
-      store.toggleFocusMode(true);
+    it('returns true if user-focused mode is configured', () => {
+      store = createStore(
+        [selection],
+        [{ focus: { user: { username: 'somebody' } } }]
+      );
 
       assert.isTrue(store.hasAppliedFilter());
```
Comment (robertknight) [inline src/sidebar/store/modules/test/selection-test.js:73]: Can you clarify the user-facing effect of the changes here?

File: src/sidebar/store/modules/test/selection-test.js [hunk 13]
```diff
@@ -107,6 +109,37 @@ describe('sidebar/store/modules/selection', () => {
     });
   });
 
+  describe('filterState', () => {
```
Comment (robertknight) [inline src/sidebar/store/modules/test/selection-test.js:112]: Can you add a test for the zero state here (ie. when no filter is active?)

File: src/styles/sidebar/components/filter-status.scss [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+  padding: var.$layout-space--small;
+  justify-content: space-between;
+
+  &__text {
+    flex-grow: 1;
+    padding-right: var.$layout-space--xsmall;
+  }
+
+  &__header {
+    // Temporary
```
Comment (robertknight) [inline src/styles/sidebar/components/filter-status.scss:36]: If this is temporary, I presume there is a plan to change this. If so, what is the eventual plan?

File: src/sidebar/store/modules/selection.js [hunk 15]
```diff
@@ -526,6 +544,25 @@ function sortKeys(state) {
   return sortKeysForTab;
 }
 
+/**
+ * Summary of applied filters
+ *
+ * @type {(state: any ) => FilterState }
```
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/sidebar/store/modules/selection.js:550]: It would be ideal to figure out what prettier/other linter setting might be allowing my editor to leave these spaces here..hrm.

File: src/sidebar/components/filter-status.js (hunk 1 above)
Comment (lyzadanger) [reply src/sidebar/components/filter-status.js:26]: Perhaps renaming this state/mode within the component would help with this confusion. Within the component, it _is_ a mutually-exclusive state, but what it means is "user-focus mode _only_". User focus mode combined with a user-entered filter would put the component in "query" mode.

That is, `focus` mode in the component does not correspond with "user-focus mode is configured" . It corresponds with "user-focus mode is configured and nothing else is going on in filter land."

File: src/sidebar/components/filter-status.js (hunk 2 above)
Comment (lyzadanger) [reply src/sidebar/components/filter-status.js:20]: Similar to my previous comment: there's a confusion here (not surprising as this stuff is complex) between this component's notion of what "focus mode" means and "user focus mode" overall.

File: src/sidebar/components/filter-status.js (hunk 1 above)
Comment (lyzadanger) [reply src/sidebar/components/filter-status.js:26]: Note: In case I forget to mention it, I've rewritten commenting in the entire module to be consistent about the use of the term "mode." _Mode_ now only refers to the three modes defined in the component itself, which are: `selection`, `query`, and `focusOnly`. I renamed the latter mode for clarity, as user-focus might be active in the other two modes, but the `focusOnly` mode indicates that it is the _only_ applied filter.

I've expanded the comments at the top of the component to try to be extra clear here. Thanks for flagging this up. It is/was indeed quite confusing.

File: src/sidebar/services/threads.js (hunk 9 above)
Comment (lyzadanger) [reply src/sidebar/services/threads.js:20]: Nope! It makes calculating the actual filter/search result count versus the forced-expanded count much easier.

In the past, the "count" of search results was simply the number of visible threads (i.e. `threads` in `rootThread` with `visible`: `true`). That meant that when a user force-expanded threads, the result count would appear to change.

It's a simple calculation to subtract the number of forced-visible threads from the total number of visible threads to get an accurate result count, but only if there's no duplication between the two sets (set A: `threads` with `visible: true`, set B: `thread`s that have been set to forced-visible).

A robust solution could be had by actually comparing items in `forcedVisible` when counting visible threads, but that was more code and more calculation, so I took this elementary approach for now.

File: src/styles/sidebar/components/filter-status.scss (hunk 14 above)
Comment (lyzadanger) [reply src/styles/sidebar/components/filter-status.scss:36]: 🤦 Whoops. Good catch.

File: src/sidebar/store/modules/selection.js (hunk 11 above)
Comment (lyzadanger) [reply src/sidebar/store/modules/selection.js:182]: I have a little bit of refactoring to do in this module as a follow up (e.g. `clearSelection` and `clearSelectedAnnotations` actions now do the same, so one should be eliminated). I'll put this on the list—cool?

File: src/sidebar/store/modules/test/selection-test.js (hunk 12 above)
Comment (lyzadanger) [reply src/sidebar/store/modules/test/selection-test.js:73]: Congratulations on finding a small bug for me! This is an artifact of mid-implementation and should be reverted. I shall do so.

File: src/sidebar/components/search-input.js (hunk 4 above)
Comment (lyzadanger) [reply src/sidebar/components/search-input.js:36]: This change has two effects:

1. Setting a search query resets the `expanded` selection state. This has the effect of folding up expanded threads as soon as a user clicks into the search input because before this change, the form would immediately submit. That's annoying: if you've got a bunch of open nested replies, they close up again as soon as you click into the input. This has been the situation for a while, and so a change for this reason _could_ be separated from this PR.
2. Setting a search query first clears the selection (including selected annotations). That means if you are viewing a set of selected annotations and then click the search icon (before these changes, when that would cause an immediate submission), all of your selected annotations are immediately cleared, even though you haven't searched for anything yet. This is relevant to these changes: this behavior didn't used to manifest, as filter queries and selected annotations would combine.

Comment (lyzadanger) [reply src/sidebar/components/search-input.js:36]: Addendum: the real fix here is that the search icon would not double as a submit button.
